### ISSUE TITLE
[java] UnusedPrivateMethod - false positive with static methods in core JDK classes

### ISSUE THREAD

**Issue by adangel** [pmd/pmd#4861]
**Affects PMD Version:** 7.0.0

**Rule:** [UnusedPrivateMethod](https://docs.pmd-code.org/latest/pmd_rules_java_bestpractices.html#unusedprivatemethod)

**Description:**

The regression tester shows, that the rule finds the following allegedly unused private methods:

- https://github.com/openjdk/jdk/tree/jdk-11+28/src/java.base/share/classes/java/lang/Integer.java#L170 'toStringUTF16(int, int)'
- https://github.com/openjdk/jdk/tree/jdk-11+28/src/java.base/share/classes/java/lang/Integer.java#L338 'toUnsignedString0(int, int)'
- https://github.com/openjdk/jdk/tree/jdk-11+28/src/java.base/share/classes/java/lang/Integer.java#L389 'formatUnsignedIntUTF16(int, int, byte[], int, int)'
- https://github.com/openjdk/jdk/tree/jdk-11+28/src/java.base/share/classes/java/lang/Long.java#L154 'toStringUTF16(long, int)'
- https://github.com/openjdk/jdk/tree/jdk-11+28/src/java.base/share/classes/java/lang/Long.java#L241 'toUnsignedBigInteger(long)'
- https://github.com/openjdk/jdk/tree/jdk-11+28/src/java.base/share/classes/java/lang/Long.java#L417 'formatUnsignedLong0UTF16(long, int, byte[], int, int)'
- https://github.com/openjdk/jdk/tree/jdk-11+28/src/java.base/share/classes/java/lang/Object.java#L41 'registerNatives()'

**Expected outcome:**

All these private static methods are indeed in use. So, none of them should be reported.

**Additional context:**

* https://github.com/pmd/pmd/pull/4841#issuecomment-1992238323
* https://github.com/pmd/pmd/pull/4841#issuecomment-1997663702
  > A very preliminar look into the FP on openjdk shows that the node is null not because inference failed, but because, as those classes (ie: `java.lang.Integer`) are part of the auxclasspath, they resolve to an ASM-backed symbol instead of an AST-backed symbol… and ASM-backed symbols can't reference the AST, so `tryGetNode()` returns `null`.
  >
  > 3 possible ways around this would be:
  > - Understand why ASM is prefered over AST-backed and see if it makes sense to switch over (probably there are performance issues at play, and ASM is objectively more reliable with things such as extending a class in a library).
  > - Change the rule to compare matches by `JMethodSig` instead of by node (this is in practice similar to what this PR tried, but without reimplementing arg-matching manually).

**Comment by adangel** [comment]
Since all these methods are in JDK classes and you mention ASM vs. AST - it could be, that the classfiles don't match exactly the source files. With the regression tester, we don't explicitly add the java runtime path to the auxclasspath, so we rely on running PMD with the correct java version...

I think, we use java 11 (latest update) when we run the regression tester, we analyze [jdk-11+28](https://github.com/openjdk/jdk/tree/jdk-11+28). The latest update would be in [jdk11u](https://github.com/openjdk/jdk11u), which should match the runtime. But I didn't see obvious changes in Integer.java.

**Comment by cowwoc** [comment]
@adangel I am getting the same false-positive on a non-JDK class.

In my case, I am getting `[INFO] PMD Failure: com.github.cowwoc.lumina.Resource:513 Rule:UnusedPrivateMethod Priority:3 Avoid unused private methods such as 'containsLink(ObjectNode)'..` on a class that looks like this:

```
public final class Resource
{
  public Link getResource(String relation) throws NoSuchRelationException
  {
    Link match = getOptionalResource(relation).orElse(null);
    [...]
  }

  private static Optional<Link> getOptionalResource(ObjectNode fromObject, String relation,	boolean insideStateMetadata)
  {
		for (Entry<String, JsonNode> entry : fromObject.properties())
		{
			String propertyName = entry.getKey();
			JsonNode propertyValue = entry.getValue();

			Optional<Link> match = switch (propertyValue)
			{
				case TextNode textNode ->
				{
					yield [...]
				}
				case ArrayNode arrayNode ->
				{
					yield [...]
				}
				case ObjectNode objectNode ->
				{
					if (containsLink(objectNode))
						yield [...]
					yield [...]
				}
				default -> Optional.empty();
			};
			if (match.isPresent())
				return match;
		}
		return Optional.empty();
  }

  private static boolean containsLink(ObjectNode object)
  {
    [...]
  }
}
```

I hope this helps.

**Comment by DashwoodIce9** [comment]
Can reproduce this issue with the following minimal code snippet -
```java
// Class1.java

package prakhar;

public class Class1 {

    public void caller() {
        Class2.class2Function(this::class1Function);
    }

    private Void class1Function() {
        System.out.println("Hello World");
        return null;
    }
}

```

```java
// Class2.java

package prakhar;

import java.util.function.Supplier;

public final class Class2 {

    private Class2() {
    }

    public static void class2Function(Supplier<Void> function) {
        function.get();
    }
}
```

`class1Function` is reported as `UnusedPrivateMethod`. Interestingly, if I move the code in just one class like so -

```java
package prakhar;

import java.util.function.Supplier;

public class Class1 {

    public static void class2Function(Supplier<Void> function) {
        function.get();
    }

    public void caller() {
        class2Function(this::class1Function);
    }

    private Void class1Function() {
        System.out.println("Hello World");
        return null;
    }
}
```

Then the violation is not reported.

**Comment by oowekyala** [comment]
This has been fixed by #5341
